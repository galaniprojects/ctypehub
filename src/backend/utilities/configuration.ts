import { cwd } from 'node:process';
import path from 'node:path';

import dotenv from 'dotenv';

dotenv.config();
const { env } = process;

const httpAuthPassword = env.SECRET_HTTP_AUTH_PASSWORD;

export const configuration = {
  port: env.PORT || 3000,
  host: env.HOST || '127.0.0.1',
  isProduction: env.NODE_ENV === 'production',
  distFolder: path.join(cwd(), 'dist', 'frontend'),
  httpAuthPassword,
  databaseUri:
    env.DATABASE_URI || 'postgres://postgres:postgres@localhost:5432/postgres',
};
