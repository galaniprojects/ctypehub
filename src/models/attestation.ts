import { type IAttestation, type ICType } from '@kiltprotocol/sdk-js';
import {
  DataTypes,
  Model,
  type ModelAttributes,
  type Sequelize,
} from 'sequelize';

export interface AttestationData
  extends Omit<IAttestation, 'cTypeHash' | 'revoked'> {
  cTypeId: ICType['$id'];
  createdAt: Date;
  extrinsicHash: string;
  block: string;
}

export class Attestation extends Model<AttestationData> {}

export const AttestationModelDefinition: ModelAttributes = {
  claimHash: {
    type: DataTypes.STRING,
    primaryKey: true,
  },
  cTypeId: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  owner: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  delegationId: {
    type: DataTypes.STRING,
  },
  createdAt: {
    type: DataTypes.DATE,
    allowNull: false,
  },
  extrinsicHash: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  block: {
    type: DataTypes.STRING,
  },
};

export function initAttestation(sequelize: Sequelize) {
  Attestation.init(AttestationModelDefinition, {
    sequelize,
    indexes: [{ fields: ['cTypeId'] }],
  });
}
