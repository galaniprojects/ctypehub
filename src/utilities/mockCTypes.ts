import type { CTypeData } from '../models/ctype';

export const mockCTypes: Record<string, CTypeData> = {
  example: {
    id: 'kilt:ctype:0x1',
    schema: 'http://kilt-protocol.org/draft-01/ctype#',
    title: 'Example CType',
    properties: { example: { type: 'string' }, isExample: { type: 'boolean' } },
    type: 'object',
    creator: 'did:kilt:4pehddkhEanexVTTzWAtrrfo2R7xPnePpuiJLC7shQU894aY',
    createdAt: new Date('2023-05-01T12:00:00'),
    extrinsicHash: '0xexamplehash',
    description: 'This is some example cType data',
    block: '123',
    attestationsCount: '1',
  },
  nestedProperty: {
    id: 'kilt:ctype:0x2',
    schema: 'http://kilt-protocol.org/draft-01/ctype#',
    title: 'Example nested CType property',
    properties: {
      example: { $ref: 'kilt:ctype:0x1#/properties/example' },
      hasNestedProperty: { type: 'boolean' },
    },
    type: 'object',
    creator: 'did:kilt:4pehddkhEanexVTTzWAtrrfo2R7xPnePpuiJLC7shQU894aY',
    createdAt: new Date('2023-05-01T12:01:00'),
    extrinsicHash: '0xexamplenestedhash',
    description: 'This is an example of a CType with a nested property',
    block: '321',
    attestationsCount: '22',
  },
  nestedCType: {
    id: 'kilt:ctype:0x3',
    schema: 'http://kilt-protocol.org/draft-01/ctype#',
    title: 'Example nested CType',
    properties: {
      example: {
        $ref: 'kilt:ctype:0x1',
      },
      hasNestedCType: {
        type: 'boolean',
      },
    },
    type: 'object',
    creator: 'did:kilt:4pehddkhEanexVTTzWAtrrfo2R7xPnePpuiJLC7shQU894aY',
    createdAt: new Date('2023-05-01T12:02:00'),
    extrinsicHash: '0xexamplehash',
    description: 'This is an example of a CType with a nested CType',
    block: '456',
    attestationsCount: '333',
  },
  everything: {
    id: 'kilt:ctype:0x060fe62fbe19e99f3440d829dec705f92d35b833d8c8927d8d94462737af2d90',
    schema:
      'ipfs://bafybeiah66wbkhqbqn7idkostj2iqyan2tstc4tpqt65udlhimd7hcxjyq/',
    title: 'Everything everywhere',
    description: 'â€¦all the time',
    additionalProperties: false,
    properties: {
      Array: {
        items: {
          format: 'date',
          type: 'string',
        },
        type: 'array',
      },
      Boolean: {
        type: 'boolean',
      },
      Integer: {
        enum: [1, 10000, 2, 3, 4],
        type: 'integer',
      },
      Number: {
        maximum: 56.78,
        minimum: 12.34,
        type: 'number',
      },
      Reference: {
        $ref: 'kilt:ctype:0xa8e68ac37f1a062b1432840d883e4c8be71deca13d539e0b00a8a74f4e3f6789',
      },
      String: {
        maxLength: 100,
        minLength: 1,
        type: 'string',
      },
    },
    type: 'object',
    creator: 'did:kilt:4rrkiRTZgsgxjJDFkLsivqqKTqdUTuxKk3FX3mKFAeMxsR5E',
    attestationsCount: '4444',
    extrinsicHash: '0xexamplehash',
    createdAt: new Date('2023-05-01T12:03:00'),
    block: '456',
  },
};
