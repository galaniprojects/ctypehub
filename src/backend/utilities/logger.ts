import { LoggerOptions, pino } from 'pino';
import hapiPino from 'hapi-pino';

import { configuration } from './configuration';

const { isProduction } = configuration;

const options: LoggerOptions = {
  level: isProduction ? 'debug' : 'trace',
  ...(!isProduction && { transport: { target: 'pino-pretty' } }),
  redact: isProduction
    ? [
        'req.headers.authorization',
        'req.headers["x-forwarded-for"]',
        'req.headers["x-real-ip"]',
      ]
    : { paths: ['req', 'res'], remove: true },
};

export const logger = pino(options);

export const hapiLogger = {
  plugin: hapiPino,
  options: {
    ...options,
    ignoreTags: ['noLogs'],
    logRequestComplete: isProduction,
  },
};
