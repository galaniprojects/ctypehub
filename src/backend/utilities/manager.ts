import { server as hapiServer } from '@hapi/hapi';
import { createManager } from 'exiting';

import { configuration } from './configuration';

const { isProduction, port, host } = configuration;

export const server = hapiServer({
  port,
  host,
  debug: isProduction ? false : undefined,
  routes: { security: true },
});

export const manager = createManager(server);
