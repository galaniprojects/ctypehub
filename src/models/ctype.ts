import { type DidUri, type ICType } from '@kiltprotocol/sdk-js';
import { DataTypes, Model, type ModelAttributes, Sequelize } from 'sequelize';

import { Attestation } from './attestation';
import { Tag } from './tag';

interface CTypeDataInput extends Omit<ICType, '$id' | '$schema'> {
  id: ICType['$id'];
  schema: ICType['$schema'];
  creator: DidUri;
  createdAt: Date;
  extrinsicHash: string;
  block: string | null;
  description: string | null;
}

export interface CTypeData extends CTypeDataInput {
  attestationsCount: string;
  isHidden: boolean;
  tags?: Pick<Tag, 'dataValues'>[];
}

export class CType extends Model<CTypeData, CTypeDataInput> {}

export const CTypeModelDefinition: ModelAttributes = {
  id: {
    type: DataTypes.STRING,
    primaryKey: true,
  },
  schema: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  title: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  properties: {
    type: DataTypes.JSON,
    allowNull: false,
  },
  type: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  creator: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  createdAt: {
    type: DataTypes.DATE,
    allowNull: false,
  },
  extrinsicHash: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  block: {
    type: DataTypes.STRING,
  },
  description: {
    type: DataTypes.STRING,
  },
  isHidden: {
    type: DataTypes.BOOLEAN,
    allowNull: false,
    defaultValue: false,
  },
};

const fields = Object.keys(CTypeModelDefinition)
  .filter((name) => name !== 'createdAt')
  .map((name) => `coalesce("${name}"::text, '')`)
  .join(` || ' ' || `);

CTypeModelDefinition.search = {
  type: `tsvector generated always as (to_tsvector('english', ${fields})) stored`,
  set() {
    throw new Error('search is read-only');
  },
};

export function initCType(sequelize: Sequelize) {
  CType.init(CTypeModelDefinition, {
    sequelize,

    scopes: {
      stats: {
        attributes: [
          // Unfortunately all the CType modelâ€™s fields have to be listed explicitly
          ...Object.keys(CTypeModelDefinition).filter(
            (key) => key !== 'search',
          ),
          [
            Sequelize.literal(
              `coalesce(
                (select count(*)
                from "Attestations"
                where "Attestations"."cTypeId" = "CType"."id"
                group by "CType"."id"),
                0)`,
            ),
            'attestationsCount',
          ],
        ],
      },
    },
  });

  CType.hasMany(Attestation, { foreignKey: 'cTypeId' });
  Attestation.belongsTo(CType, { foreignKey: 'cTypeId' });

  CType.hasMany(Tag, { foreignKey: 'cTypeId', as: 'tags' });
  Tag.belongsTo(CType, { foreignKey: 'cTypeId' });
}
