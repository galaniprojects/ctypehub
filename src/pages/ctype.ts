import { type APIContext } from 'astro';
import { CType, Did, type DidUri, type HexString } from '@kiltprotocol/sdk-js';
import { StatusCodes } from 'http-status-codes';

import { CType as CTypeModel } from '../models/ctype';
import { generatePath, paths } from '../paths';
import { type TagDataInput, Tag as TagModel } from '../models/tag';
import {
  handleErrorResponse,
  makeErrorResponse,
} from '../utilities/errorResponses';
import { getRequestJson } from '../utilities/getRequestJson';

interface Input {
  cType: unknown;
  creator: DidUri;
  extrinsicHash: HexString;
  description: string;
  tags: string[];
}

export async function POST({ request, url }: APIContext) {
  try {
    const { cType, creator, extrinsicHash, description, tags } =
      await getRequestJson<Input>(request);

    if (!CType.isICType(cType)) {
      throw makeErrorResponse(
        'Only ICType is accepted',
        StatusCodes.UNPROCESSABLE_ENTITY,
      );
    }

    try {
      Did.validateUri(creator, 'Did');
    } catch {
      throw makeErrorResponse(
        'Creator is not a valid DID',
        StatusCodes.UNPROCESSABLE_ENTITY,
      );
    }

    const { $id: id, $schema: schema, ...rest } = cType;
    if (await CTypeModel.findByPk(id)) {
      throw makeErrorResponse(
        'This CType already exists',
        StatusCodes.CONFLICT,
      );
    }

    await CTypeModel.create({
      id,
      schema,
      creator,
      extrinsicHash,
      description,
      createdAt: new Date(),
      ...rest,
    });

    const cTypeTags: TagDataInput[] = tags.map((tag) => ({
      cTypeId: id,
      tagName: tag,
    }));
    await TagModel.bulkCreate(cTypeTags);

    const cTypeUrl = new URL(url);
    cTypeUrl.pathname = generatePath(paths.ctypeDetails, id);

    return new Response('', {
      status: StatusCodes.CREATED,
      headers: {
        Location: cTypeUrl.toString(),
      },
    });
  } catch (exception) {
    return handleErrorResponse(exception);
  }
}
