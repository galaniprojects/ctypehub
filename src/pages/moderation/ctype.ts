import type { APIContext } from 'astro';

import { StatusCodes } from 'http-status-codes';
import { Op } from 'sequelize';

import { getSession } from '../session';
import {
  handleErrorResponse,
  makeErrorResponse,
} from '../../utilities/errorResponses';
import { CType } from '../../models/ctype';

export async function GET({ request, url }: APIContext) {
  try {
    const session = getSession(request);
    if (!session.authorized) {
      throw makeErrorResponse('Unauthorized', StatusCodes.UNAUTHORIZED);
    }

    const before = url.searchParams.get('before');

    const cTypes = await CType.findAll({
      where: {
        createdAt: { [Op.lte]: before ? new Date(before) : new Date() },
      },
      order: [['createdAt', 'DESC']],
      include: 'tags',
      limit: 20,
    });

    return new Response(JSON.stringify(cTypes), {
      headers: { 'Content-Type': 'application/json' },
    });
  } catch (exception) {
    return handleErrorResponse(exception);
  }
}
