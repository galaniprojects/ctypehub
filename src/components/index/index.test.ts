import { beforeEach, describe, expect, it } from '@jest/globals';

import { getSnapshotHtmlForPath } from '../../../testing/getSnapshotHtmlForPath';
import { CType as CTypeModel } from '../../models/ctype';
import { mockCTypes } from '../../utilities/mockCTypes';

beforeEach(async () => {
  await CTypeModel.destroy({ truncate: true });
  await CTypeModel.bulkCreate([mockCTypes.example, mockCTypes.nestedProperty]);
});

describe('index.astro', () => {
  it('should render all CTypes', async () => {
    const html = await getSnapshotHtmlForPath('');
    expect(html).toMatchSnapshot();
  });

  it('should render search results', async () => {
    const html = await getSnapshotHtmlForPath('?query=nested');
    expect(html).toMatchSnapshot();
  });
  it('should render when no search results', async () => {
    const html = await getSnapshotHtmlForPath('?query=asdagalasdg');
    expect(html).toMatchSnapshot();
  });
});
