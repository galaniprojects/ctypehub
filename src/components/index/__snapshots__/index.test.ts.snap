// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`index.astro > should render all CTypes except hidden ones 1`] = `
<body>
  <header>
    <a href="/" class="logo"><img src="/src/components/Header/logo.svg" alt="CTypeHub logo" /></a>
    <a href="/create" class="_primary"> Create Claim Type </a>
  </header>
  <main>
    <search>
      <form action="/" method="get" class="_container">
        <h1>CType Hub</h1>
        <div>
          <input type="search" name="query" placeholder="Search query" aria-label="Search query" class="_search1 _component" /><button type="submit" class="_search8 _primary">Search</button>
        </div>
      </form>

      <ul>
        <li class="_container container">
          <h2>
            <a href="/ctype/kilt:ctype:0x2" class="title">
              Example nested CType property
            </a>
          </h2>

          <p class="description">
            This is an example of a CType with a nested property
          </p>

          <p class="attestations">
            <span>Number of Attestations:</span>
            <span>2</span>
          </p>
        </li>
        <li class="_container container">
          <h2>
            <a href="/ctype/kilt:ctype:0x1" class="title"> Example CType </a>
          </h2>

          <p class="description">This is some example cType data</p>

          <ul class="tags">
            <li class="tag">
              <a href="/tag/example">#example</a>
            </li>
          </ul>

          <p class="attestations">
            <span>Number of Attestations:</span>
            <span>1</span>
          </p>
        </li>
      </ul>
    </search>
  </main>
  <footer>
    <nav>
      <!-- TODO: links to legal docs -->
    </nav>
    <p>
      Made with ♥︎ by
      <img class="galaniLogo" src="/src/components/Footer/galaniprojects.svg" alt="galaniprojects logo" />
      © 2023
    </p>
    <img class="kiltLogo" src="/src/components/Footer/built_on_kilt.svg" alt="built on KILT logo" />
  </footer>
</body>
`;

exports[`index.astro > should render search results 1`] = `
<body>
  <header>
    <a href="/" class="logo"><img src="/src/components/Header/logo.svg" alt="CTypeHub logo" /></a>
    <a href="/create" class="_primary"> Create Claim Type </a>
  </header>
  <main>
    <search>
      <form action="/" method="get" class="_container">
        <h1>CType Hub</h1>
        <div>
          <input type="search" name="query" value="nested" placeholder="Search query" aria-label="Search query" class="_search1 _component" /><button type="submit" class="_search8 _primary">Search</button>
        </div>
      </form>

      <p>CType results for “nested”:</p>
      <ul>
        <li class="_container container">
          <h2>
            <a href="/ctype/kilt:ctype:0x2" class="title">
              Example nested CType property
            </a>
          </h2>

          <p class="description">
            This is an example of a CType with a nested property
          </p>

          <p class="attestations">
            <span>Number of Attestations:</span>
            <span>2</span>
          </p>
        </li>
      </ul>
    </search>
  </main>
  <footer>
    <nav>
      <!-- TODO: links to legal docs -->
    </nav>
    <p>
      Made with ♥︎ by
      <img class="galaniLogo" src="/src/components/Footer/galaniprojects.svg" alt="galaniprojects logo" />
      © 2023
    </p>
    <img class="kiltLogo" src="/src/components/Footer/built_on_kilt.svg" alt="built on KILT logo" />
  </footer>
</body>
`;

exports[`index.astro > should render tag results 1`] = `
<body>
  <header>
    <a href="/" class="logo"><img src="/src/components/Header/logo.svg" alt="CTypeHub logo" /></a>
    <a href="/create" class="_primary"> Create Claim Type </a>
  </header>
  <main>
    <search>
      <form action="/" method="get" class="_container">
        <h1>CType Hub</h1>
        <div>
          <input type="search" name="query" value="example" placeholder="Search query" aria-label="Search query" class="_search1 _component" /><button type="submit" class="_search8 _primary">Search</button>
        </div>
      </form>

      <p>Tag results for “#example”:</p>

      <ul class="tagResults">
        <li>
          <a class="_container" href="/tag/example"> #example </a>
        </li>
      </ul>

      <p>CType results for “example”:</p>
      <ul>
        <li class="_container container">
          <h2>
            <a href="/ctype/kilt:ctype:0x2" class="title">
              Example nested CType property
            </a>
          </h2>

          <p class="description">
            This is an example of a CType with a nested property
          </p>

          <p class="attestations">
            <span>Number of Attestations:</span>
            <span>2</span>
          </p>
        </li>
        <li class="_container container">
          <h2>
            <a href="/ctype/kilt:ctype:0x1" class="title"> Example CType </a>
          </h2>

          <p class="description">This is some example cType data</p>

          <ul class="tags">
            <li class="tag">
              <a href="/tag/example">#example</a>
            </li>
          </ul>

          <p class="attestations">
            <span>Number of Attestations:</span>
            <span>1</span>
          </p>
        </li>
      </ul>
    </search>
  </main>
  <footer>
    <nav>
      <!-- TODO: links to legal docs -->
    </nav>
    <p>
      Made with ♥︎ by
      <img class="galaniLogo" src="/src/components/Footer/galaniprojects.svg" alt="galaniprojects logo" />
      © 2023
    </p>
    <img class="kiltLogo" src="/src/components/Footer/built_on_kilt.svg" alt="built on KILT logo" />
  </footer>
</body>
`;

exports[`index.astro > should render the tag page 1`] = `
<body>
  <header>
    <a href="/" class="logo"><img src="/src/components/Header/logo.svg" alt="CTypeHub logo" /></a>
    <a href="/create" class="_primary"> Create Claim Type </a>
  </header>
  <main>
    <search>
      <form action="/tag/example" method="get" class="_container">
        <h1>#example</h1>
        <div>
          <input type="search" name="query" placeholder="Search query" aria-label="Search query" class="_search1 _component" /><button type="submit" class="_search8 _primary">Search</button>
        </div>
      </form>

      <ul>
        <li class="_container container">
          <h2>
            <a href="/ctype/kilt:ctype:0x1" class="title"> Example CType </a>
          </h2>

          <p class="description">This is some example cType data</p>

          <ul class="tags">
            <li class="tag">
              <a href="/tag/example">#example</a>
            </li>
          </ul>

          <p class="attestations">
            <span>Number of Attestations:</span>
            <span>1</span>
          </p>
        </li>
      </ul>
    </search>
  </main>
  <footer>
    <nav>
      <!-- TODO: links to legal docs -->
    </nav>
    <p>
      Made with ♥︎ by
      <img class="galaniLogo" src="/src/components/Footer/galaniprojects.svg" alt="galaniprojects logo" />
      © 2023
    </p>
    <img class="kiltLogo" src="/src/components/Footer/built_on_kilt.svg" alt="built on KILT logo" />
  </footer>
</body>
`;

exports[`index.astro > should render when no search results 1`] = `
<body>
  <header>
    <a href="/" class="logo"><img src="/src/components/Header/logo.svg" alt="CTypeHub logo" /></a>
    <a href="/create" class="_primary"> Create Claim Type </a>
  </header>
  <main>
    <search>
      <form action="/" method="get" class="_container">
        <h1>CType Hub</h1>
        <div>
          <input type="search" name="query" value="asdagalasdg" placeholder="Search query" aria-label="Search query" class="_search1 _component" /><button type="submit" class="_search8 _primary">Search</button>
        </div>
      </form>

      <p>No CType results for “asdagalasdg”</p>
    </search>
  </main>
  <footer>
    <nav>
      <!-- TODO: links to legal docs -->
    </nav>
    <p>
      Made with ♥︎ by
      <img class="galaniLogo" src="/src/components/Footer/galaniprojects.svg" alt="galaniprojects logo" />
      © 2023
    </p>
    <img class="kiltLogo" src="/src/components/Footer/built_on_kilt.svg" alt="built on KILT logo" />
  </footer>
</body>
`;
